elytron/WFCORE-4360-CredentialStore_Expression_Resolution.adoc

```diff
@@ -0,0 +1,420 @@
+operations which will become complex very quickly a new set of actions will be added:
+
+ * `--generate-secret-key`
+ * `--export-secret-key`
+ * `--import-secret-key`  
+
+The `--generate-secret-key` action will take the alias to store the key as an argument in the same pattern as `add` it will also take the following additional
+option:
+
+ * `--size` - The key size (number of bits) - acceptable values 128, 192, or 256 - default 256.
+ 
+The `--export-secret-key` action will take the alias of the previously stored key as an argument, no further option will be required.
+
+The `--export-secret-key` action will return a single Base64 value which is the key in it's encoded form.
+
+The `--import-secret-key` action will take the alias to store the key as an argument, it will also take the following additional option: -
+
+ * `--key` - The key to import encoded using Base64.
+
+If the `--key` option is not specified the user will be prompted to enter the key interactively, this will avoid the key being seen in the current running processes and avoid the key being cached in the users command line history.
+
+NOTE: For `generate-secret-key` and `import-secret-key` if an entry already exists for that alias / algorithm combination the behaviour will be the same as the `add` action in relation to replacing the entry or reporting an error.  (To be checked)
+
+For this specific enhancement where an algorithm is required for the `SecretKey` we will use the hard coded value of `AES`, if later enhancement add support for further key types then it may be necessary for an optional `algorithm` argument / parameter to be added to each of the above commands / operations.  Should an `algorithm` argument be added at a later point it would be optional and default to `AES` for backwards compatibility.
+
+It is not feasible for all actions to have a short form so these new actions will have a long form only, within the tool we should try and restrict the use of the short form for parameters only.
+
+==== Key manipulation - Management Operations
+
+A set of operations for secret key manipulation will be added to both the existing `credential-store` resource and the new `secret-key-credential-store`
+resource.  Presently the `credential-store` resource contains an `add-alias` operation which can be used to add a credential based on a clear text String,
+the new operations will perform a similar purpose.
+
+The following operations will be added to the `credential-store` resource:
+
+ * `generate-secret-key`
+ * `export-secret-key`
+ * `import-secret-key`
+ 
+Within the management model operations are self describing so making use of dedicated operations makes it easier for tooling to present a meaningful UI to end
+users automatically.
+
+The `credential-store` resource already contains a `remove-alias` operation however this assumes the type of the credential is `PasswordCredential` a new
```

Comment:
There is also the `read-aliases` operation. The `entry-type` parameter should be available also in this case, or the output should be structured to cover all the entry types.

Reply:
For now the credential store already supports two different entry types and this missing information is already an issue in the present implementation, this RFE is adding support for a third entry type.

I agree we should improve our read-aliases support but I think it should be a part of a small follow up RFE as it ends up with it's own set of compatibility issues.

Firstly the credential store APIs do not expose this information so changes to the SPI could impact existing implementations.

I would suggest the read-aliases operation should have a new parameter "show-type" to switch on showing type information, there should also be a second paramter "types" so the caller can specify the types to return in the output.

I don't mind getting this raised into a new RFE unless you really want to push that this should be solved for WildFly 23.

Reply:
I have no strong arguments against moving this to a follow-up. Can create a new RFE for this please?

Reply:
I will create that now then we can close this discussion.

Reply:
Here is a new RFE to handle this https://issues.redhat.com/browse/EAP7-1643